src/main/java/org/opentripplanner/updater/trip/TimetableSnapshotSource.java

```diff
@@ -690,12 +699,92 @@ private Result<?, UpdateError> handleAddedTrip(
     return addTripToGraphAndBuffer(
       tripBuilder.build(),
       tripUpdate,
+      stopTimeUpdates,
       stops,
       serviceDate,
       RealTimeState.ADDED
     );
   }
 
+  private Route getOrCreateRoute(TripDescriptor tripDescriptor, FeedScopedId tripId) {
+    // Create dummy agency for added trips
+    Agency fallbackAgency = Agency
+      .of(new FeedScopedId(tripId.getFeedId(), "autogenerated-gtfs-rt-added-route"))
+      .withName("Agency automatically added by GTFS-RT update")
+      .withTimezone(transitService.getTimeZone().toString())
+      .build();
+
+    Route route;
+
+    // this route in this update doesn't already exist so it will be created
+    if (
+      tripDescriptor.hasExtension(MfdzRealtimeExtensions.tripDescriptor) &&
+      !routeExists(tripId.getFeedId(), tripDescriptor)
+    ) {
+      FeedScopedId id = new FeedScopedId(tripId.getFeedId(), tripDescriptor.getRouteId());
+
+      var builder = Route.of(id);
+
+      var addedRouteExtension = AddedRoute.ofTripDescriptor(tripDescriptor);
+
+      var agency = transitService
+        .getAgencies()
+        .stream()
+        .filter(a -> a.getId().getFeedId().equals(tripId.getFeedId()))
+        .filter(addedRouteExtension::matchesAgencyId)
+        .findFirst()
+        .orElse(fallbackAgency);
+
+      builder.withAgency(agency);
+
+      builder.withGtfsType(addedRouteExtension.routeType());
+      var mode = TransitModeMapper.mapMode(addedRouteExtension.routeType());
+      builder.withMode(mode);
+
+      // Create route name
+      var name = Objects.requireNonNullElse(addedRouteExtension.routeLongName(), tripId.toString());
+      builder.withLongName(new NonLocalizedString(name));
+      builder.withShortName(name);
+      builder.withUrl(addedRouteExtension.routeUrl());
+      builder.withAgency(fallbackAgency);
+
+      route = builder.build();
+
+      transitService.addRoutes(route);
+    } else if (routeExists(tripId.getFeedId(), tripDescriptor)) {
+      // Try to find route
+      route =
+        transitService.getRouteForId(
+          new FeedScopedId(tripId.getFeedId(), tripDescriptor.getRouteId())
+        );
```

Comment:
We should probably have this at the top of the method and return early, in order to avoid creating unnecessary `fallbackAgency`

Reply:
I've reworked the generation in 693f0c4b3dac22b1c29defa6fab12772bd34cef1.

Reply:
I think we could still improve this by lifting the case where the route already exists to the top, and use early return in it. The two other branches share some code, so they could be partially merged. Also we don't need to have the `route` local variable, as it is only used for storing it once and returning immediately.

Reply:
I reworked again in https://github.com/opentripplanner/OpenTripPlanner/commit/f7a983b7db86c8b08f84b04e08b5161eeeb3d63c